rmgpy/solver/base.pyx

```diff
@@ -359,7 +359,39 @@ cdef class ReactionSystem(DASx):
             for l, spec in enumerate(network.source):
                 i = self.get_species_index(spec)
                 self.networkIndices[j,l] = i
-
+                                   
+    @cython.boundscheck(False)                               
+    cpdef maxIndUnderSurfaceLayeringConstraint(self,numpy.ndarray[numpy.float64_t,ndim=1] arr,list surfSpeciesIndices):
+        """
+        determines the index under the surface layering constraint, all of the reactants or all of the products must be in the
+        bulk core (in the core, but not in the surface)
+        """
+        
+        cdef numpy.ndarray[numpy.int_t, ndim=1] sortedIndices
+        cdef int i,j,k,index, numCoreSpecies
+        cdef bool boo
+        
+        numCoreSpecies = self.numCoreSpecies
+        productIndices= self.productIndices
+        reactantIndices = self.reactantIndices
+        
+        sortedIndices = getReverseSortedIndices(arr)
+        
+        for i in xrange(len(sortedIndices)):
+            index = sortedIndices[i]
+            boo = True
+            for j in productIndices[index]:
+                if j in surfSpeciesIndices or j > numCoreSpecies:
+                    boo = False
+            if boo:
+                return index
+            boo = True
+            for j in reactantIndices[index]:
+                if j in surfSpeciesIndices or j > numCoreSpecies:
+                    boo = False
+            if boo:
+                return index
+            
```

Comment:
I think there are a few differences, although it certainly should go over sortedIndices and the for else structure is certainly much cleaner, although I personally don't like them very much.  The indices in sortedIndices (index) are edge reaction indices, j indexes species.  

Since they are edge indices you have to add numCoreSpecies to go over the right species indices in productIndices and reactantIndices and it is >= since j == numCoreSpecies will be in the edge.